### PR TITLE
[Backport 2.x] Initial search pipelines implementation (#6587)

### PR DESCRIPTION
**Pull request by msfroh**
* Initial search pipelines implementation

This commit includes the basic features of search pipelines (see https://github.com/opensearch-project/search-processor/issues/80).

Search pipelines are modeled after ingest pipelines and provide a simple, clean API for components to modify search requests and responses.

With this commit we can:

1. Can create, retrieve, update, and delete search pipelines.
2. Transform search requests and responses by explicitly referencing a pipeline.

Later work will include:

1. Adding an index setting to specify a default search pipeline.
2. Allowing search pipelines to be defined within a search request (for development/testing purposes, akin to simulating an ingest pipeline).
3. Adding a collection of search pipeline processors to support common useful transformations. (Suggestions welcome!)

Signed-off-by: Michael Froh <froh@amazon.com>

* Incorporate feedback from @reta and @navneet1v

1. SearchPipelinesClient: JavaDoc fix
2. SearchRequest: Check versions when (de)serializing new "pipeline" property.
3. Rename SearchPipelinesPlugin -> SearchPipelinePlugin.
4. Pipeline: Change visibility to package private
5. SearchPipelineProcessingException: New exception type to wrap exceptions thrown when executing a pipeline.

Bonus: Added an integration test for filter_query request processor.

Signed-off-by: Michael Froh <froh@amazon.com>

* Register SearchPipelineProcessingException

Also added more useful messages to unit tests to explicitly explain what hoops need to be jumped through in order to add a new serializable exception.

Signed-off-by: Michael Froh <froh@amazon.com>

* Remove unneeded dependencies from search-pipeline-common

I had copied some dependencies from ingest-common, but they are not used by search-pipeline-common (yet).

Signed-off-by: Michael Froh <froh@amazon.com>

* Avoid cloning SearchRequest if no SearchRequestProcessors

Also, add tests to confirm that a pipeline with no processors works fine (as a no-op).

Signed-off-by: Michael Froh <froh@amazon.com>

* Use NamedWritableRegistry to deserialize SearchRequest

Queries are serialized as NamedWritables, so we need to use a NamedWritableRegistry to deserialize.

Signed-off-by: Michael Froh <froh@amazon.com>

* Check for empty pipeline with CollectionUtils.isEmpty

Signed-off-by: Michael Froh <froh@amazon.com>

* Update server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java

Co-authored-by: Navneet Verma <vermanavneet003@gmail.com>
Signed-off-by: Michael Froh <froh@amazon.com>

* Incorporate feedback from @noCharger

Signed-off-by: Michael Froh <froh@amazon.com>

* Incorporate feedback from @reta

- Renamed various classes from "SearchPipelinesSomething" to "SearchPipelineSomething" to be consistent.
- Refactored NodeInfo construction in NodeService to avoid ternary operator and improved readability.

Signed-off-by: Michael Froh <froh@amazon.com>

* Gate search pipelines behind a feature flag

Also renamed SearchPipelinesRequestConverters.

Signed-off-by: Michael Froh <froh@amazon.com>

* More feature flag fixes for search pipeline testing

- Don't use system properties for SearchPipelineServiceTests.
- Enable feature flag for multinode smoke tests.

Signed-off-by: Michael Froh <froh@amazon.com>

* Move feature flag into constructor parameter

Thanks for the suggestion, @reta!

Signed-off-by: Michael Froh <froh@amazon.com>

* Move REST handlers behind feature flag

Signed-off-by: Michael Froh <froh@amazon.com>

---------

Signed-off-by: Michael Froh <froh@amazon.com>
Co-authored-by: Navneet Verma <vermanavneet003@gmail.com>
(cherry picked from commit ee990bd40ceacbf9ebd6ddea0874aa98c48ece47)

### CODE REVIEW

**Review by reta — APPROVED** [review]

File: server/src/main/java/org/opensearch/action/admin/cluster/node/info/NodeInfo.java [hunk 1]
```diff
@@ -102,6 +103,9 @@ public NodeInfo(StreamInput in) throws IOException {
         if (in.getVersion().onOrAfter(LegacyESVersion.V_7_10_0)) {
             addInfoIfNonNull(AggregationInfo.class, in.readOptionalWriteable(AggregationInfo::new));
         }
+        if (in.getVersion().onOrAfter(Version.V_2_7_0)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/action/admin/cluster/node/info/NodeInfo.java:106]: @msfroh we would need to bring these changes to `main` as well, could you please prepare the pull request so we will merge it right after this one? thank you

**Comment by msfroh** [comment]
I'm canceling this PR for now, just to make sure it doesn't go out in its current state for 2.7. I'll see about reopening after 2.7 is released.

**Comment by msfroh** [comment]
Talked it over with a colleague, who mentioned that the feature flag is enough that it should be safe to ship this code in 2.7.

File: server/src/main/java/org/opensearch/action/admin/cluster/node/info/NodeInfo.java (hunk 1 above)
Comment (msfroh) [reply server/src/main/java/org/opensearch/action/admin/cluster/node/info/NodeInfo.java:106]: Done, thanks! Here we go: https://github.com/opensearch-project/OpenSearch/pull/7135

**Comment by reta** [comment]
@andrross just to confirm, good to go for 2.7.0? thank you

**Comment by saratvemulapalli** [comment]
> Talked it over with a colleague, who mentioned that the feature flag is enough that it should be safe to ship this code in 2.7.

Late here. Im the mystery colleague. As long as we dont break semver and the feature is gated we should be good.

**Pull request merged**